docs/design/logging-integration.adoc

```diff
@@ -0,0 +1,196 @@
+- When the `Pod` is archived, the URL for accessing the log should be generated
+and stored with it. There is no attempt to query the logging system to verify
+the existence of the log.
+
+====
+
+Here's another sample `KubeArchiveConfig` for `PipelineRuns`:
+```
+---
+apiVersion: kubearchive.kubearchive.org/v1alpha1
+kind: KubeArchiveConfig
+metadata:
+  name: kubearchive
+  namespace: test
+spec:
+  resources:
+    - selector:
+        apiVersion: tekton.dev/v1
+        kind: PipelineRun
+      deleteWhen: has(status.completionTime)
+    - selector:
+        apiVersion: tekton.dev/v1
+        kind: TaskRun
+      archiveOnDelete: true
+    - selector:
+        apiVersion: v1
+        kind: Pod
+      archiveOnDelete: has(body.metadata.labels["tekton.dev/pipeline"])
+```
+In this case the following happens:
+
+- `PipelineRuns` are archived when they complete.
+- `TaskRuns` are archived when they are deleted.
+- `Pods` are archived when they are deleted and are also part of a Tekton Pipeline
+
+== Generating Log URLs
+
+The logging system is an integration in KubeArchive. All that is required for
+the integration is the successful generation of a URL to access the log for a
+specific `Pod` in the logging system. URLs will ONLY be generated for `Pods`.
+
+To support multiple logging systems, the URLs must be able to be parameterized
+based on the logging system. This could be done via a `ConfigMap` named for the
+logging system being implemented. This `ConfigMap` requires a single entry,
+LOG_URL, whose string value will be evaluated in an environment where variables
+in it will be substituted. For example, take this `ConfigMap` for Splunk:
+```
+---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: kubearchive-splunk
+  namespace: kubearchive
+data:
+  LOG_URL=https://splunk-host/app/search/search?q=search *| spath "kubernetes.pod_id" | search "kubernetes.pod_id"="${POD_ID}"
+  POD_ID=".body.pod_id"
+```
+Any other entry in this `ConfigMap` is treated as a variable, and the value of
+each variable is a JSON path expression that is evaluated against the body of
+the cloud event (i.e. the resource) to determine the real value of that variable
+used in the substitution. For example:
+
+So when generating the logging URL to be stored when a `Pod` is archived,
+the following steps are done:
+
+1. An "environment" is created by evaluating the JSON path expression in each entry
+   in the `ConfigMap`, excluding LOG_URL.
+1. The value for LOG_URL is then evaluated using this environment, resulting in the
+   final URL to the log in the logging system.
+
+== Implementation
+
+=== Database
+
+The KubeArchive database will have a table named `log_url` with three fields:
+
+1. An `id` field as a primary key.
+1. A `resource_id` field which is a foreign key to `resource.id`.  
+1. A `url` field which is the URL for one of of logs.
+
+The `resource_id` field should point back to a `Pod` entry in the `resource` table.
```

Comment:
Shouldn't we also include a way to provide the relationship of "owner-owned" among the different resources and their pods? For example, between a `Deployment` and their `Pods`, between a `CronJob` and their `Pods` between a `PipelineRun` and their `Pods`.

Consider that currently with `kubectl` we can run `kubectl logs deploy/<my-deploy>` and retrieve the logs of all the pods that are currently running (or that ran in the past) this deployment. I guess that KubeArchive should provide a pointer to the same logs based on the same parameters.

The deployment and their pods are related based on a labelselector. We currently have all the information in the db, we have the labelSelector of a deployment and the labels of the pods but that query might be tricky, so I suggest adding an additional table to relate these objects more easily.

Reply:
> I'm not sure how your comment relates to the log_url table. That table is simply a list of log URLs that tie back to pods in the resource table.

Yes, sorry for introducing confusion, my concern here is about the need of an extra table, it's not related to the `log_url` table :)

> This may be true, but they are also related by "owner-owned". But there is a ReplicaSet in between, so that will have to be archived as well. 

Ok so we have the `ownerReference` in the `metadata` of the different resources, and based on that we can go from a pod to an owner (like a `Deployment`) directly or indirectly (in the case of the `Deployment` we need to go through the `ReplicaSet` as you mentioned). 
But my point here is that the user may want to provide the name of a `Deployment` to get the logs of its "owned" pods.
The problem I see is going from the `Deployment` to the `Pods` and not the other way around. We currently have some information in the database that allow us to do that, the `labelSelector` of the `Deployment` and the `labels` in the `metadata` of the `Pods` but I believe that will be somewhat unefficient.

So my suggestion was to think of a new relational database to store the `owner` -> `owned` relationship instead to be able to extract the log references based on the name of an "owner"

Reply:
@maruiz93, the JSON indexes should allow for easy retrieval of the owner relationship data without the need for an extra table.  These queries are very fast.

Yes, you may have to make multiple queries to get the entire list of logs.  But it is doable. But one thing to consider is possible writing a stored procedure to do this work.

Reply:
For future reference, I also created an issue with a suggested stored procedure for doing the work of relate a resource with its owned `Pods` to test it in the future in case this could improve the performance: https://github.com/kubearchive/kubearchive/issues/492

Reply:
This is an implementation decision that should be made when the implementation is done.  Both should be tried.  I think the stored procedure might be simpler than doing it in code.